src/nvim/ops.c: Fix bug

Zman94 opened the pull request:
Fix bug in do_join.

justinmk commented on src/nvim/ops.c:
```diff
@@ -3644,6 +3644,10 @@ int do_join(size_t count,
    */
   for (t = (linenr_T)count - 1;; t--) {
     cend -= currsize;
+	if(cend == NULL || curr == NULL){
```
`cend` was dereferenced on line 3636, so it can't be NULL here.

AFAICT `curr` cannot be NULL unless count=0. And the docstring for the function claims that `count` minimum is 2. 

So I would the only thing needed is 

    assert(cound >= 2); 

near the start of the function.

Zman94 replied:
I'll do that, but I don't see where `curr` depends on `count`. Additionally, we keep subtracting from `cend`, shouldn't we still check for NULL as we subtract?

justinmk replied:
I don't know. Try the first thing first...

Zman94 replied:
As I was trying this out, `exec make -C build unittest` closes my terminal. Ever encounter this issue before? I retested on a vanilla copy of the code and ran into the same issue.

jamessan replied:
@Zman94 That part of the documentation is referring to `git rebase -i`.  `exec` is one of the recognized keywords for the rebase buffer which tells git to run a command (in this case `make -C build unittest`) before moving on to the next commit.

If you instead ran this manually from your shell, then it's expected that the terminal exited.  The `exec` builtin in the shell replaces the shell process with the subsequent command.  Therefore, when make is done running, there's nothing running in the terminal anymore.

Zman94 replied:
Ok. So just to clarify, those rebase commands that execute the unit tests are run as a branch is rebased or merged back into master automatically?

jamessan replied:
The intention of the instructions is to ensure the individual commits are still sane when doing a rebase, so that we don't break the ability to perform bisects later on.

So, if you're doing a rebase it's a good idea to insert `exec make -C build unittest` lines after every commit that's being kept.  Then when git processes the rebase, it will run `make -C build unittest` after every rebased commit.

Zman94 replied:
How is that done? `git rebase <feature branch> -i exec` while in master? Additionally, why is rebase used instead of merging? Is it just preference?

jamessan replied:
Please see [CONTRIBUTING.md](https://github.com/neovim/neovim/blob/master/CONTRIBUTING.md) for explanations about that.

However, given that you didn't have any conflicts with master, I'm not sure why you were trying a rebase or merge.  If there's no need to do either of those, then don't bother.

Zman94 replied:
Yes. I've read it a few times; I'm just new to contributing to open source and trying to figure out how. Please excuse some of the more ignorant questions.

I didn't encounter any conflicts. I wrote my code in a feature branch then merged it into master then created a pull request. I'm just trying to figure out why my build was failing on unit tests, but it seems as though I shouldn't have been running that.

Zman94 replied:
Let me clarify that. I'm unsure what I need to do to get my changes merged into master.

jamessan replied:
There's no need to merge to your local master branch.  Work is typically done on your computer in a distinct feature branch.  That feature branch is pushed to Github.  When/if the project maintainers decide to include the change, they will merge it into master (or whatever the target branch is).

Zman94 replied:
Ok. Thank you for the help. I was already in master branch on my local computer because I thought I needed to merge the files first. Thank you for the clarification. I just pushed a new update, hopefully you can see it.

justinmk commented:
Try `make lint` locally also :)

zachary-gleason committed:
ops.c: do_join expects `count` of 2 or greater #6855

test_normal.vim fails if we assert >= 2, so we only check >= 1. That at
least avoids invalid pointers.

    TEST_FILE=test_normal.res make oldtest

Zman94 commented:
Just want to bump this. Any thoughts on the last push?

Zman94 commented:
"You don't need to open a new PR just because the old one hasn't been merged yet." - Sorry @jamessan , I thought that because it was a new bug, it needed a new PR.

"Note that this PR is failing the same way the last PR was:
`src/nvim/ops.c:3644: int do_join(size_t, int, int, int, _Bool): Assertion `count >= 2' failed.
Aborted (core dumped)
make: *** [test_normal.res] Error 134`
That needs to be fixed before this can be merged." - I'm a little unsure where you found that error. Could you tell me the commands needed to get that error? I'll fix it then resubmit a new comment.

"Either do that here and close #6855 or close this and update #6855." - Closing #6912

jamessan commented:
> I'm a little unsure where you found that error.

It's in the log for the failed Travis CI builds.

> Could you tell me the commands needed to get that error?

Run `make oldtest` to run the legacy Vim tests.

Zman94 commented:
@jamessan I'm looking through the tests to understand why my code failed the legacy tests. What language are the files in src/nvim/testdir that end in .vim? Additionally, does this seem like a productive approach to debugging?

jamessan commented:
> What language are the files in src/nvim/testdir that end in .vim?

They're vim script.  Each function is a set of commands that get run in nvim and calls to the `assert_*` family of functions to check the results are as expected.

> Additionally, does this seem like a productive approach to debugging?

Yes, understanding the test that's hitting the assertion is a good idea.  You can run just the `test_normal.vim` tests with the command `make -B -C src/nvim/testdir test_normal.res`.  The test results will be appended to `src/nvim/testdir/messages`, so after the abort is fixed this file should also show that all tests pass.

Zman94 commented:
Hello, I'm struggling with the issues left on this pull request. I can't seem to understand the Test_normal33_g_cmd2() line 79 where it says norm! 10gJ. It then asserts 1 is equal to col('.') which I believe is the current line. I'm also not too sure on what I'm skipping over in the do_join function within src/nvim/ops.c

justinmk commented:
`col('.')` is the current column ("." means "current thing" where "thing" depends on the context).

> I'm also not too sure on what I'm skipping over in the do_join function within src/nvim/ops.c

Not sure what you mean. Could you describe the bug you originally were intending to fix?

It looks like my [suggestion](https://github.com/neovim/neovim/pull/6855#discussion_r120649121) was not quite right, so something else is needed.

Zman94 commented:
Yeah. I changed your assert to an if statement that returns failure. I was looking through the do_join function to see what it was that wasn't executing because that if statement returns. I'll take another look tonight.

Zman94 commented:
I just went through it again, and it seems as if I end up at the first column. I'm using debug.vim, a file I created that only holds the function Test_normal33_g_cmd2(). I'll push the changes I have.

Zman94 commented:
I'm confused by what is actually broken. When I run `make -B -C src/nvim/testdir test_normal.res` I don't have any errors, but I have a few tests that were skipped. When I run `make oldtest`, I get this error:

`From test_cscope.vim:
Found errors in Test_cscopeWithCscopeConnections():
function RunTheTest[9]..Test_cscopeWithCscopeConnections line 154: Pattern '(1 of 1): <<index_to_key>> #define index_to_key(i) ((i) ^ 15167)' does not match '"samples/memfile_test.c" 146L, 3303C'
function RunTheTest[9]..Test_cscopeWithCscopeConnections line 161: command did not fail: tag TEST_COUNT
function RunTheTest[9]..Test_cscopeWithCscopeConnections line 164: Pattern '(1 of 1): <<TEST_COUNT>> #define TEST_COUNT 50000' does not match ''
function RunTheTest[9]..Test_cscopeWithCscopeConnections line 167: Pattern '1  1 TEST_COUNT\s\+\d\+\s\+#define index_to_key' does not match '  # TO tag         FROM line  in file/text  1  1 index_to_key       29  Xmemfile_test.c  2  1 TEST_COUNT         28  #define index_to_key(i) ((i) ^ 15167)  3  1 TEST_COUNT         29  #define TEST_COUNT 50000>'
TEST FAILURE
`

I also got this error when I ran make oldtest without my changes in. Is this bug because of my code?

jamessan commented:
> I also got this error when I ran make oldtest without my changes in. Is this bug because of my code?

Since it happens without your changes, it can't be because of your code.  Also, the cscope code has nothing to do with `do_join`.

What's the actual bug you're trying to fix here?

justinmk commented:
Yeah, ignore `Test_cscopeWithCscopeConnections` test failure. It fails for me locally, too. It could depend on the version of cscope your system has, etc. (Though it would be great to fix this)

> What's the actual bug you're trying to fix here?

I'm also wondering as well. What are the steps a user can take to reproduce the bug? If we don't have any steps, then we can't write a test...

jamessan commented:
My best guess is that it's the scan-build [error](https://neovim.io/doc/reports/clang/report-6c4a6e.html#EndPath) that `curr` could be NULL in the

```c
memmove(cend, curr, (size_t)currsize);
```

however that can only occur when `do_join` is called with a `count` of 0.  That's invalid according to the documented contract of the function, so I think `do_join` should simply `assert(count > 0)` upon entry.  That should be enough to satisfy scan-build.

justinmk commented:
ah, maybe that's better than my previous suggestion https://github.com/neovim/neovim/pull/6855/commits/f951e7c39a87f2cbba0a30e88f8239fa43208171

Zman94 commented:
The bug I'm trying to fix is why my push is failing the CI tests.

Additionally, I've not had any issues with `curr` being equal to 0 as far as I can tell in my logs.

justinmk commented:
Pushed an update to this PR. Looks like `[OLDTEST] test_normal` fails when `count` is asserted at the start of `do_join`.
